### PR TITLE
added primary changes for synthesis

### PR DESCRIPTION
**Pull request by abhigya-sodani**

### CODE REVIEW

**Comment by abhigya-sodani** [comment]
Basically added a function that creates (prompt,response) pairs based on the given prompt and the "best" response for it (this is the completion we get from our "big" model). This function asks an LLM to create JSON objects "prompt" and "response" feilds, and future work will largely center around making sure that this process is robust.

**Comment by cartazio** [comment]
One concern i have that maybe doesn’t matter: will the contents of those json fields be properly escaped? Json is a format where parenthesis matching and quote escaping etc can be kinda important.

**Comment by cartazio** [comment]
I guess I mean : whatever encoding we ask an llm to do, there should be a way to avoid ever interpreting the output as being unparsable? Or maybe that’s a feature? I really don’t have a good intuition around this stuff.

File: src/llm_vm/completion/optimize.py [hunk 1]
```diff
@@ -216,14 +215,29 @@ def complete_delay_train(self, stable_context, dynamic_prompt, c_id = None, **kw
...
+                        data = '{ "datapoints": ['+data+']}'
+                        print(data)
+                        data = json.loads(data)
+                        for i in data["datapoints"]:
+                            datapoints.append((i["prompt"],i["response"]))
+                        return datapoints
+
                     train_completion = best_completion if use_completion is None else use_completion
-                    new_datapoint = (dynamic_prompt.strip(), train_completion)
+                    new_datapoint = (prompt.strip(), train_completion)
```
Comment (mmirman) [inline src/llm_vm/completion/optimize.py:230]: Why did you change dynamic_prompt to prompt?   It was originally dynamic_prompt so if the static_prompt was huge the smaller model wouldn't get it in the context window and it'd be even more optimized

File: src/llm_vm/completion/optimize.py [hunk 2]
```diff
@@ -216,14 +215,29 @@ def complete_delay_train(self, stable_context, dynamic_prompt, c_id = None, **kw
             def promiseCompletion():
                 best_completion = self.call_big(prompt, **kwargs)
                 def actual_train(use_completion = None):
+                    def data_synthesis(prompt, response, **kwargs):
+                        final_prompt = '{"prompt": "' +prompt+'"  , "response": "' +response+'" }'+'\nGenerate 100 more JSONS each with a prompt and response field like the given one. The content of the prompt and response fields must be similar to the given JSON. Seperate each JSON with a ,.'
```
Comment (mmirman) [inline src/llm_vm/completion/optimize.py:219]: I'm a bit concerned that the statement "similar to the given JSON" will cause too much repetition.

File: src/llm_vm/completion/optimize.py [hunk 3]
```diff
@@ -216,14 +215,29 @@ def complete_delay_train(self, stable_context, dynamic_prompt, c_id = None, **kw
...
+                        for i in data["datapoints"]:
+                            datapoints.append((i["prompt"],i["response"]))
+                        return datapoints
+
                     train_completion = best_completion if use_completion is None else use_completion
-                    new_datapoint = (dynamic_prompt.strip(), train_completion)
+                    new_datapoint = (prompt.strip(), train_completion)
                     self.storage.add_example(c_id, new_datapoint)
+                    for j in data_synthesis(prompt,best_completion):
+                        self.storage.add_example(c_id, j)
```
Comment (mmirman) [inline src/llm_vm/completion/optimize.py:233]: @abhigya-sodani Won't this lead to polynomial growth of the data in storage synthesis?

File: src/llm_vm/completion/optimize.py (hunk 2 above)
Comment (mmirman) [reply src/llm_vm/completion/optimize.py:219]: Also, 100 should probably be a variable modifiable by the object

Comment (unna97) [reply src/llm_vm/completion/optimize.py:219]: When running this as a code this currently generated the task by just changing the country.Couple ways I thought of solving this problem:
1. Separate into "prompt template" (a GPT-3 recognised word) and variables. Then ask the gpt-3 to change specific paramaters for the template.
        Prompt template: What is the [variable] of the [object]? 
                        - variable-object pairs = 1. currency, Myanmar
                                                  2. price, Bitcoin
Now, for generation we might specify differences for either prompt templates or variable/object.Based on following parameters(not comprehensive list):
  i. Style: Keep the meaning of the question same but ask in different styles.
            eg: introduce spelling mistakes, code mixing, cultural tone difference etc
  ii. Semantic diversity: The task itself should change.

2. Add the above nlp metrics in code and set thresholds, keep generating till the thersholds are satisfied.
    Reference: 1. https://arxiv.org/pdf/2005.04118.pdf (behavioral  testing of the nlp)
               2.https://arxiv.org/abs/2111.09963 (behavioral testing of recommnder system)
   essentially we will framing the task as NLP task or recommendation task with behaviour metrics satistfied.
   eg: keep generating similar sentences with cosine_similarity 0.5 and diversity 0.4 till "n" sentences reached or "x" attempts satisfied.

File: src/llm_vm/completion/optimize.py [hunk 4]
```diff
@@ -216,14 +215,29 @@ def complete_delay_train(self, stable_context, dynamic_prompt, c_id = None, **kw
             def promiseCompletion():
                 best_completion = self.call_big(prompt, **kwargs)
                 def actual_train(use_completion = None):
+                    def data_synthesis(prompt, response, **kwargs):
+                        final_prompt = '{"prompt": "' +prompt+'"  , "response": "' +response+'" }'+'\nGenerate 100 more JSONS each with a prompt and response field like the given one. The content of the prompt and response fields must be similar to the given JSON. Seperate each JSON with a ,.'
+                        data=self.call_big(final_prompt , kwargs)
+                        datapoints = []
+                        data = '{ "datapoints": ['+data+']}'
```
Comment (unna97) [inline src/llm_vm/completion/optimize.py:222]: Maybe the process of converting the data to datapoints i.e. Getting the synthesized data from the response could be a separate function that can be called here?

File: src/llm_vm/completion/optimize.py (hunk 3 above)
Comment (unna97) [reply src/llm_vm/completion/optimize.py:233]: I do think it's important to assign a variable to generated dataset, have some basic checks before adding it to storage?

File: src/llm_vm/completion/optimize.py (hunk 1 above)
Comment (abhigya-sodani) [reply src/llm_vm/completion/optimize.py:230]: ok will change back, I probably got a bit confused by the way it was being used in test_optimize

File: src/llm_vm/completion/optimize.py (hunk 2 above)
Comment (abhigya-sodani) [reply src/llm_vm/completion/optimize.py:219]: > I'm a bit concerned that the statement "similar to the given JSON" will cause too much repetition.

As in not different enough examples, and fine tuning on too similar data examples?

Comment (abhigya-sodani) [reply src/llm_vm/completion/optimize.py:219]: > eep the meaning of the question same but ask in different styles.
>                    eg: introduce spelling mistakes, code mixing, cultural tone difference etc.
> 
>        ii. Semantic diversity: The task itself should chang

This seems like a really good idea to try, I will make some tickets

File: src/llm_vm/completion/optimize.py (hunk 3 above)
Comment (abhigya-sodani) [reply src/llm_vm/completion/optimize.py:233]: > @abhigya-sodani Won't this lead to polynomial growth of the data in storage synthesis?

I am just adding all generated examples for finetuning, do we want to clear out the storage after finetuning is done to prevent the polynomial growth?
